hw/drivers/uart/max3107/src/max3107.c

```diff
@@ -0,0 +1,1059 @@
+                if (abs((int)(actual_br - br)) < abs((int)(best_br - br))) {
+                    best_br = actual_br;
+                    cfg->div_lsb = (uint8_t)(div >> 4);
+                    cfg->div_msb = (uint8_t)(div >> 12);
+                    cfg->brg_config = (div & 0xF) | ((mul == 4) ? 0x20 : ((mul == 2) ? 0x10 : 0));
+                    /* If best choice is factor_ix == 0, no need for PLL */
+                    if (factor_ix == 0) {
+                        cfg->clk_source |= CLCSOURCE_PLLBYPASS;
+                        cfg->clk_source &= ~CLCSOURCE_PLLEN;
+                    } else {
+                        cfg->pll_config = pre_div | ((factor_ix - 1) << 6);
+                        cfg->clk_source &= ~CLCSOURCE_PLLBYPASS;
+                        cfg->clk_source |= CLCSOURCE_PLLEN;
+                    }
+                }
+            }
+        }
+    }
+
+    return best_br;
+}
+
+int
+max3107_config_uart(struct max3107_dev *dev, const struct uart_conf_port *conf)
+{
+    int rc;
+
+    if (dev->cfg.crystal_en) {
+        dev->regs.clock.clk_source |= CLCSOURCE_CRYSTALEN;
+    } else {
+        dev->regs.clock.clk_source &= ~CLCSOURCE_CRYSTALEN;
+    }
+    if (dev->cfg.no_pll) {
+        dev->regs.clock.clk_source |= CLCSOURCE_PLLBYPASS;
+    } else {
+        dev->regs.clock.clk_source &= ~CLCSOURCE_PLLBYPASS;
+    }
+    if (dev->cfg.allow_mul_4) {
+        dev->regs.clock.brg_config = BRGCONFIG_4XMODE;
+    } else if (dev->cfg.allow_mul_2) {
+        dev->regs.clock.brg_config = BRGCONFIG_2XMODE;
+    } else {
+        dev->regs.clock.brg_config = 0;
+    }
+    max3107_compute_clock_config(dev->cfg.osc_freq, conf->uc_speed, &dev->regs.clock);
+
+    rc = max3107_write_regs(dev, MAX3107_REG_PLLCONFIG, &dev->regs.clock.pll_config, 5);
+    if (rc) {
+        goto end;
+    }
+
+    dev->regs.modes.mode1 = MODE1_IRQSEL | (conf->uc_flow_ctl ? 0 : MODE1_RTSHIZ);
+    dev->regs.modes.mode2 = 0;
+    dev->regs.modes.lcr = (dev->uart.ud_conf_port.uc_stopbits == 1 ? 0 : LCR_STOPBITS) |
+                          (dev->uart.ud_conf_port.uc_databits - 5) |
+                          ((dev->uart.ud_conf_port.uc_parity != UART_PARITY_NONE) ? LCR_PARITYEN : 0 ) |
+                          ((dev->uart.ud_conf_port.uc_parity == UART_PARITY_EVEN) ? LCR_EVENPARITY : 0);
+    dev->regs.modes.rxtimeout = 2;
+    dev->regs.modes.hdplxdelay = 0;
+    dev->regs.modes.irda = 0;
+
+    rc = max3107_write_regs(dev, MAX3107_REG_MODE1, &dev->regs.modes.mode1, 6);
+    if (rc) {
+        goto end;
+    }
+
+    /* RTS activated at FIFO level 120. */
+    dev->regs.fifo.flow_lvl = 0xFF;
+    dev->regs.fifo.fifo_trg_lvl = 0x38;
+    rc = max3107_write_regs(dev, MAX3107_REG_FLOWLVL, &dev->regs.fifo.flow_lvl, 2);
+    if (rc) {
+        goto end;
+    }
+
+    dev->regs.flow.flow_ctrl = conf->uc_flow_ctl ? (FLOWCTRL_AUTOCTS | FLOWCTRL_AUTORTS) : 0;
+    max3107_write_reg(dev, MAX3107_REG_FLOWCTRL, dev->regs.flow.flow_ctrl);
```

Comment:
We need to handle error here.

Reply:
All cases are now handled.
Client will be notified about catastrophic failures in which case closing and opening device should be done.